clients/drcachesim/tests/instr_decode_cache_test.cpp

```diff
@@ -0,0 +1,167 @@
+    }
+    bool is_nop = false;
+    bool is_ret = false;
+};
+
+std::string
+check_decode_caching_without_instr()
+{
+    static constexpr addr_t BASE_ADDR = 0x123450;
+    static constexpr addr_t TID_A = 1;
+    instr_t *nop = XINST_CREATE_nop(GLOBAL_DCONTEXT);
+    instr_t *ret = XINST_CREATE_return(GLOBAL_DCONTEXT);
+    instrlist_t *ilist = instrlist_create(GLOBAL_DCONTEXT);
+    instrlist_append(ilist, nop);
+    instrlist_append(ilist, ret);
+    std::vector<memref_with_IR_t> memref_setup = {
+        { gen_instr(TID_A), nop },
+        { gen_instr(TID_A), ret },
+        { gen_instr(TID_A), nop },
+    };
+    auto memrefs = add_encodings_to_memrefs(ilist, memref_setup, BASE_ADDR);
+    // Set up the second nop memref to reuse the same encoding as the first nop.
+    memrefs[2].instr.encoding_is_new = false;
+
+    instr_decode_cache_t<test_decode_info_t> decode_cache(
+        GLOBAL_DCONTEXT,
+        /*persist_decoded_instrs=*/false);
+    if (decode_cache.get_decode_info(reinterpret_cast<app_pc>(memrefs[0].instr.addr)) !=
+        nullptr) {
+        return "Unexpected decode info for never-seen pc";
+    }
+    decode_cache.add_decode_info(memrefs[0].instr);
+    test_decode_info_t *decode_info_nop =
+        decode_cache.get_decode_info(reinterpret_cast<app_pc>(memrefs[0].instr.addr));
+    if (decode_info_nop == nullptr || !decode_info_nop->is_nop) {
+        return "Unexpected decode info for nop instr";
+    }
+    decode_cache.add_decode_info(memrefs[1].instr);
+    test_decode_info_t *decode_info_ret =
+        decode_cache.get_decode_info(reinterpret_cast<app_pc>(memrefs[1].instr.addr));
+    if (decode_info_ret == nullptr || !decode_info_ret->is_ret) {
+        return "Unexpected decode info for ret instr";
+    }
+    decode_cache.add_decode_info(memrefs[2].instr);
+    test_decode_info_t *decode_info_nop_2 =
+        decode_cache.get_decode_info(reinterpret_cast<app_pc>(memrefs[2].instr.addr));
+    if (decode_info_nop_2 != decode_info_nop) {
+        return "Did not see same decode info instance for second instance of nop";
+    }
+
+    instrlist_clear_and_destroy(GLOBAL_DCONTEXT, ilist);
+    std::cerr << "check_decode_caching_without_instr passed\n";
+    return "";
+}
+
+std::string
+check_instr_decode_caching()
+{
+    static constexpr addr_t BASE_ADDR = 0x123450;
+    static constexpr addr_t TID_A = 1;
+    instr_t *nop = XINST_CREATE_nop(GLOBAL_DCONTEXT);
+    instr_t *ret = XINST_CREATE_return(GLOBAL_DCONTEXT);
+    instrlist_t *ilist = instrlist_create(GLOBAL_DCONTEXT);
+    instrlist_append(ilist, nop);
+    instrlist_append(ilist, ret);
+    std::vector<memref_with_IR_t> memref_setup = {
+        { gen_instr(TID_A), nop },
+        { gen_instr(TID_A), ret },
+        { gen_instr(TID_A), nop },
+    };
+    auto memrefs = add_encodings_to_memrefs(ilist, memref_setup, BASE_ADDR);
+    // Set up the second nop memref to reuse the same encoding as the first nop.
+    memrefs[2].instr.encoding_is_new = false;
```

Comment:
Could we share the setup code?  Put this inside the other function with {} scopes for the two tests?

Reply:
Done.